drafts/element-internals-api.md

```diff
@@ -0,0 +1,179 @@
+properties from our native checkbox on our component.
+
+```typescript
+import {customElement, property} from 'lit/decorators.js';
+import {LitElement} from 'lit';
+import {createRef, Ref, ref} from 'lit-html/directives/ref.js';
+
+@customElement('dss-switch')
+export default class Switch extends LitElement {
+  static formAssociated = true;
+
+  @property({type: Boolean})
+  public checked = false;
+
+  @property()
+  public value?: string;
+
+  private internals: ElementInternals;
+  private inputRef: Ref<HTMLInputElement> = createRef();
+
+  constructor() {
+    super();
+    this.internals = this.attachInternals();
+  }
+
+  protected render() {
+    return html`
+      <input 
+        ${ref(this.inputRef)}
+        type="checkbox"
+        ?checked="${this.checked}"
+        @change=${(event: Event) => this.handleCheckboxChange(event)}
+      >
+    `;
+  }
+
+  private handleCheckboxChange(event: Event) {
+    const checkbox = event.target as HTMLInputElement;
+    this.checked = checkbox.checked;
+    this.value = checkbox.value;
+    this.dispatchEvent(new Event('change', event));
+  }
+
+  protected updated(changedProperties: PropertyValues): void {
+    if (changedProperties.has('checked')) {
+      this.internals.setFormValue(this.checked ? 'on' : null);
+      this.internals.setValidity(
+        this.inputRef.value!.validity,
+        this.inputRef.value!.validationMessage,
+        this.inputRef!.value,
+      );
+    }
+    super.updated(changedProperties);
+  }
+}
+```
+
+Let's break this example down. First, we render the native input and pass the property `checked`
+down to it. This means, when we set that property on our web component, it will update the native
+checkbox.
+
+Next, we set up an event listener for the change event on the `checkbox`, which will trigger a
+handler function. Within this function, we will modify the `checked` and `value` properties of our
+component based on the changes made to the native checkbox. Since change events do not travel
+through the ShadowDOM we have to re-dispatch the event.
+
+Lastly, whenever our components `checked` property changes we will update the ElementInternals. The
+native checkbox has `null` or `on` set as the value, so we replicate this behavior in our component.
+Updating the validity does not make a lot of sense yet, but later if we would support things
+like `required` we could take the native checkbox validation messages. This means when trying to
+submit the form our component would report the validity of the native checkbox.
+
+## Focus events
+
+Many 3rd party form libraries allow you to run validations on `change` or on `blur`. This means our
+web component needs to dispatch those events. For our `switch` component, we already dispatch
+a `change` event, and we won't have to worry about focus/blur events. Since we have a native input
+in our ShadowDOM our component will dispathc focus and blur events when the native input receives or
```

Comment:
I think there should be a comma after ShadowDOM

Reply:
Sounds right